gluten-data/src/main/scala/org/apache/gluten/metrics/InputIteratorMetricsUpdater.scala

```diff
@@ -15,22 +15,29 @@
  * limitations under the License.
  */
 package org.apache.gluten.metrics
+import org.apache.spark.TaskContext
 import org.apache.spark.sql.execution.metric.SQLMetric
 
-case class InputIteratorMetricsUpdater(metrics: Map[String, SQLMetric]) extends MetricsUpdater {
+case class InputIteratorMetricsUpdater(
+    metrics: Map[String, SQLMetric],
+    forBroadcast: Boolean = false)
+  extends MetricsUpdater {
   override def updateNativeMetrics(opMetrics: IOperatorMetrics): Unit = {
     if (opMetrics != null) {
       val operatorMetrics = opMetrics.asInstanceOf[OperatorMetrics]
       metrics("cpuCount") += operatorMetrics.cpuCount
       metrics("wallNanos") += operatorMetrics.wallNanos
-      if (operatorMetrics.outputRows == 0 && operatorMetrics.outputVectors == 0) {
-        // Sometimes, velox does not update metrics for intermediate operator,
-        // here we try to use the input metrics
-        metrics("numOutputRows") += operatorMetrics.inputRows
-        metrics("outputVectors") += operatorMetrics.inputVectors
-      } else {
-        metrics("numOutputRows") += operatorMetrics.outputRows
-        metrics("outputVectors") += operatorMetrics.outputVectors
+      // For broadcast exchange, we only collect the metrics once.
+      if (!forBroadcast || TaskContext.getPartitionId() == 0) {
+        if (operatorMetrics.outputRows == 0 && operatorMetrics.outputVectors == 0) {
+          // Sometimes, velox does not update metrics for intermediate operator,
+          // here we try to use the input metrics
+          metrics("numOutputRows") += operatorMetrics.inputRows
+          metrics("outputVectors") += operatorMetrics.inputVectors
+        } else {
+          metrics("numOutputRows") += operatorMetrics.outputRows
+          metrics("outputVectors") += operatorMetrics.outputVectors
+        }
```

Comment:
Thank you very much for working on this though I feel this solution is a little bit tricky. Could we just add a metric like `numPartitions` to `InputIteratorTransformer` which may help user understand the duplication of data?

Reply:
Hi @zhztheplayer any suggestions about how to proceed with this?

Reply:
@ivoson Sorry for the late reply. I just went through the relevant code, one of my questions is do we actually need to insert a `InputIteratorTransformer` between BHJ and BE? As BE RDD is handled differently with regular RDDs.

Can we just remove `InputIteratorTransfomer` in this case?

Reply:
> Can we just remove InputIteratorTransfomer in this case?

Thanks @zhztheplayer , let me check on this.

Reply:
Thank you cc @ulysses-you

Reply:
Remove InputIteratorTransfomer would break the assuming of operators when doing transform. It's not worthy to me to do this big change but only fix a minor issue.. Besides, even we remove InputIteratorTransfomer, I think the join itself would count input rows like InputIteratorTransfomer.

Reply:
> Remove InputIteratorTransfomer would break the assuming of operators when doing transform

@ulysses-you I was thinking about only removing the ones between BHJs and BEs. Normal `InputIteratorTransfomer`s that converts Spark iterators to Velox inputs are good and should be kept. 

Once we included broadcast RDDs into normal Substrait transformation procedure (if we can manage to do that), we can add `InputIteratorTransfomer` back then restart the discussion about #7079 again. Before that, I think `InputIteratorTransfomer`s between BHJs and BEs are misleading on UT as they don't actually do Spark-to-Substrait transformation.

> Besides, even we remove InputIteratorTransfomer, I think the join itself would count input rows like InputIteratorTransfomer.

Thought it could be a bit more acceptable since it's literally a "broadcast" join on UI.

Reply:
For velox backend, the code path of BHJ is same with SHJ, so we would go to substrait transformation code path for the BHJ InputIteratorTransfomers. The only difference thing from SHJ is, BHJ using a fake RDD which just hold a iterator.

Reply:
> Once we included broadcast RDDs into normal Substrait transformation procedure

It's my mistake here. You are right and the build side iterator is already included in Substrait plan.

Reply:
What I was worried about is when something like `TaskContext.getPartitionId() == 0)` is coded, then there might be some corner cases that the broadcasted data are not identically consumed by different join partitions. Say if probe side of one partition is an empty input, backend could do some kind of optimization that prevents the build iterator from being read. Then wrong metrics will be displayed. @ivoson @ulysses-you If you guys could help confirm.

If that's possible, we may still need a better solution to make sure we don't introduce new hidden issues.

Reply:
Yeah, I agree this may be a potential issue. But it should not only about broadcast iterator. I think we should make metrics framework work with all empty iterator case, e.g., union an empty partition and do shuffle ? Not sure if it would cause some unexpected behavior.

Reply:
> e.g., union an empty partition and do shuffle 

I assume this case is not that similar to the current one? Say we are reporting only the first partition of input iterator transformer on BE side by condition `if (TaskContext.getPartitionId() == 0)`. Thing is, partition 0 could report zero input rows while partition 1 may report non-zero rows in the corner case. cc @ivoson

Reply:
> TaskContext.getPartitionId() == 0

I see, you are reference this pr's code. I did not take deep look in this pr before... Can we pass the broadcasted row count to metrics directly ? I think we do not need to depend on the velox metrics to get the row count.

Reply:
> Can we pass the broadcasted row count to metrics directly ?

Yes, this could be a feasible solution to me too.

Reply:
Thanks @zhztheplayer @ulysses-you for the feedback, you are right about the concerns. Checking more details about other potential solutions...